pallets/blindbox/src/lib.rs

```diff
@@ -0,0 +1,288 @@
+
+    #[pallet::storage]
+    #[pallet::getter(fn is_init)]
+    pub(super) type Init<T: Config> = StorageValue<_, bool, ValueQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn nonce)]
+    pub(super) type Nonce<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub (super) fn deposit_event)]
+    #[pallet::metadata(T::AccountId = "AccountId")]
+    pub enum Event<T: Config> {
+        RandomnessConsumed(H256, H256),
+        BlindBoxIdGenerated(Vec<u32>),
+        BlindBoxOpened(T::AccountId, BlindBoxType, u32),
+        BlindBoxGoodLuckNextTime(u32),
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        //No permission
+        NoPermission,
+        // BlindBox does not exist
+        BlindBoxDoesNotExist,
+        // BlindBoxes still available, only allow to create once all the blindboxes have been used
+        BlindBoxesStillAvailable
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        #[pallet::weight(10_000)]
+        pub(super) fn set_blindbox_caller(origin: OriginFor<T>, account_id: T::AccountId ) -> DispatchResultWithPostInfo {
+            let _ = ensure_root(origin)?;
+
+            BlindBoxesCreators::<T>::insert(account_id, {});
+
+            Ok(().into())
+        }
+
+        #[pallet::weight(10_000)]
+        pub(super) fn generate_blindbox_ids(origin: OriginFor<T>, number_blindboxes: u32) -> DispatchResultWithPostInfo {
+            let caller = ensure_signed(origin)?;
+
+            // Ensure the authorized caller can call this func
+            ensure!(
+                BlindBoxesCreators::<T>::contains_key(&caller),
+                Error::<T>::NoPermission
+            );
+
+            // Ensure caller can only generate blindboxes once all the available blindboxes have been used
+            ensure!(
+                AvailableBlindBoxesCount::<T>::get() <= 0,
+                Error::<T>::BlindBoxesStillAvailable
+            );
+
+            let mut blindbox_vec = Vec::new();
+
+            // Generate random blindbox id and store
+            let mut number_blindboxes_generated = 0;
+            let mut i = 0;
+
+            // Add safe check in case of infinite loop, running extra 10 loops to generate unique blindbox id
+            while number_blindboxes_generated < number_blindboxes && i < number_blindboxes + 10 {
+                let mut blindbox_id = Self::generate_random_number(i);
+
+                if !BlindBoxes::<T>::contains_key(blindbox_id) {
+                    // Push to Vec and save to storage
+                    blindbox_vec.push(blindbox_id);
+                    BlindBoxes::<T>::insert(blindbox_id, {});
+
+                    number_blindboxes_generated += 1;
+                }
+
+                i += 1;
+            }
+
+            AvailableBlindBoxesCount::<T>::put(number_blindboxes_generated);
+            AvailableKSM::<T>::put(200000); // 20KSM
```

Comment:
This will add  AvailableKSM = 20 every time generating a new Blindbox?

Reply:
yes, or shall we have a method to allow user set it?

Reply:
I don't think 20 KSM every time generating new Blindbox. You can pass it as a Type, no need method for setting it.